File: alumnium/clients/http_client.py
from requests import delete, get, post

from ..server.agents.retriever_agent import Data
from ..server.models import Model
from ..tools.base_tool import BaseTool
from ..tools.tool_to_schema_converter import convert_tools_to_schemas


class HttpClient:
    def __init__(self, base_url: str, model: Model, platform: str, tools: dict[str, type[BaseTool]]):
        self.base_url = base_url.rstrip("/")
        self.session_id = None

        tool_schemas = convert_tools_to_schemas(tools)

        response = post(
            f"{self.base_url}/v1/sessions",
            json={
                "provider": model.provider.value,
                "name": model.name,
                "tools": tool_schemas,
                "platform": platform,
            },
            timeout=30,
        )
        response.raise_for_status()
        self.session_id = response.json()["session_id"]

    def quit(self):
        if self.session_id:
            response = delete(
                f"{self.base_url}/v1/sessions/{self.session_id}",
                timeout=30,
            )
            response.raise_for_status()
            self.session_id = None

    def plan_actions(self, goal: str, accessibility_tree: str, area_id: int = None):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/plans",
            json={
                "goal": goal,
                "accessibility_tree": accessibility_tree,
                "area_id": area_id,
            },
            timeout=120,
        )
        response.raise_for_status()
        return response.json()["steps"]

    def add_example(self, goal: str, actions: list[str]):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            json={"goal": goal, "actions": actions},
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def clear_examples(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            timeout=30,
        )
        response.raise_for_status()

    def execute_action(self, goal: str, step: str, accessibility_tree: str, area_id: int = None):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/steps",
            json={"goal": goal, "step": step, "accessibility_tree": accessibility_tree, "area_id": area_id},
            timeout=120,
        )
        response.raise_for_status()
        return response.json()["actions"]

    def retrieve(
        self,
        statement: str,
        accessibility_tree: str,
        title: str,
        url: str,
        screenshot: str | None,
        area_id: int = None,
    ) -> tuple[str, Data]:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/statements",
            json={
                "statement": statement,
                "accessibility_tree": accessibility_tree,
                "title": title,
                "url": url,
                "screenshot": screenshot if screenshot else None,
                "area_id": area_id,
            },
            timeout=120,
        )
        response.raise_for_status()
        data = response.json()
        return data["explanation"], data["result"]

    def find_area(self, description: str, accessibility_tree: str):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/areas",
            json={"description": description, "accessibility_tree": accessibility_tree},
            timeout=60,
        )
        response.raise_for_status()
        data = response.json()
        return {"id": data["id"], "explanation": data["explanation"]}

    def find_element(self, description: str, accessibility_tree: str, area_id: int = None):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/elements",
            json={"description": description, "accessibility_tree": accessibility_tree, "area_id": area_id},
            timeout=60,
        )
        response.raise_for_status()
        return response.json()["elements"][0]

    def save_cache(self):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    def discard_cache(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    @property
    def stats(self):
        response = get(
            f"{self.base_url}/v1/sessions/{self.session_id}/stats",
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

File: packages/core/src/clients/HttpClient.ts
import { AppId } from "../AppId.js";
import type { Http } from "../Http.js";
import { LlmUsageStats } from "../llm/llmSchema.js";
import { ErrorResponse } from "../server/serverSchema.js";
import { convertToolsToSchemas } from "../tools/toolToSchemaConverter.js";
import { getLogger } from "../utils/logger.js";
import type {
  AddExampleRequest,
  AreaRequest,
  AreaResponse,
  ChangesRequest,
  ChangesResponse,
  FindRequest,
  FindResponse,
  PlanRequest,
  PlanResponse,
  SessionRequest,
  SessionResponse,
  StatementRequest,
  StatementResponse,
  StepRequest,
  StepResponse,
} from "./ApiModels.js";
import { Client } from "./Client.js";
import { type Data, looselyTypecast } from "./typecasting.js";

const logger = getLogger(import.meta.url);

export namespace HttpClient {
  export interface Props extends Client.Props {
    baseUrl: string;
  }
}

export class HttpClient extends Client {
  static TIMEOUT: number = 300_000; // 5 minutes

  #baseUrl: string;
  #sessionIdPromise: Promise<string>;

  constructor(props: HttpClient.Props) {
    const { baseUrl, ...superProps } = props;
    super(superProps);
    this.#baseUrl = baseUrl.replace(/\/$/, "");
    this.#sessionIdPromise = this.#initSession();
  }

  async quit(): Promise<void> {
    await this.#sessionFetch("DELETE", "/sessions");
  }

  async planActions(
    goal: string,
    accessibilityTree: string,
    app: AppId,
  ): Promise<Client.PlanActionsResult> {
    const body: PlanRequest = {
      goal,
      accessibility_tree: accessibilityTree,
      app,
    };
    return this.#sessionFetch<PlanResponse>("POST", "/plans", body);
  }

  async addExample(goal: string, actions: string[]): Promise<void> {
    const body: AddExampleRequest = {
      goal,
      actions,
    };
    await this.#sessionFetch("POST", "/examples", body);
  }

  async clearExamples(): Promise<void> {
    await this.#sessionFetch("DELETE", "/examples");
  }

  async executeAction(
    goal: string,
    step: string,
    accessibilityTree: string,
    app: AppId,
  ): Promise<Client.ExecuteActionResult> {
    const body: StepRequest = {
      goal,
      step,
      accessibility_tree: accessibilityTree,
      app,
    };
    return this.#sessionFetch<StepResponse>("POST", "/steps", body);
  }

  async retrieve(
    statement: string,
    accessibilityTree: string,
    title: string,
    url: string,
    app: AppId,
    screenshot?: string,
  ): Promise<[string, Data]> {
    const body: StatementRequest = {
      statement,
      accessibility_tree: accessibilityTree,
      title,
      url,
      screenshot: screenshot || null,
      app,
    };
    const result = await this.#sessionFetch<StatementResponse>(
      "POST",
      "/statements",
      body,
    );
    return [result.explanation, looselyTypecast(result.result)];
  }

  async findArea(
    description: string,
    accessibilityTree: string,
    app: AppId,
  ): Promise<Client.FindAreaResult> {
    const body: AreaRequest = {
      description,
      accessibility_tree: accessibilityTree,
      app,
    };
    const data = await this.#sessionFetch<AreaResponse>("POST", "/areas", body);
    return { id: data.id, explanation: data.explanation };
  }

  async findElement(
    description: string,
    accessibilityTree: string,
    app: AppId,
  ): Promise<Client.FindElementResult | undefined> {
    const body: FindRequest = {
      description,
      accessibility_tree: accessibilityTree,
      app,
    };
    const result = await this.#sessionFetch<FindResponse>(
      "POST",
      "/elements",
      body,
    );
    return result.elements[0];
  }

  async analyzeChanges(
    beforeAccessibilityTree: string,
    beforeUrl: string,
    afterAccessibilityTree: string,
    afterUrl: string,
    app: AppId,
  ): Promise<string> {
    const body: ChangesRequest = {
      before: { accessibility_tree: beforeAccessibilityTree, url: beforeUrl },
      after: { accessibility_tree: afterAccessibilityTree, url: afterUrl },
      app,
    };
    const result = await this.#sessionFetch<ChangesResponse>(
      "POST",
      "/changes",
      body,
    );
    return result.result;
  }

  async saveCache(): Promise<void> {
    await this.#sessionFetch("POST", "/caches");
  }

  async discardCache(): Promise<void> {
    await this.#sessionFetch("DELETE", "/caches");
  }

  async getStats(): Promise<LlmUsageStats> {
    return this.#sessionFetch<LlmUsageStats>("GET", "/stats");
  }

  async #sessionFetch<Result>(
    method: Http.Method,
    path: string,
    body?: unknown,
  ): Promise<Result> {
    return this.#withSessionId((sessionId) =>
      this.#fetch(method, `/sessions/${sessionId}${path}`, body),
    );
  }

  async #withSessionId<Result>(
    fn: (sessionId: string) => Result | Promise<Result>,
  ): Promise<Result> {
    const sessionId = await this.#sessionIdPromise;
    return fn(sessionId);
  }

  async #initSession(): Promise<string> {
    const toolSchemas = convertToolsToSchemas(this.tools);
    const body: SessionRequest = {
      provider: this.model.provider,
      name: this.model.name,
      platform: this.platform as SessionRequest["platform"],
      tools: toolSchemas,
      planner: this.planner,
      exclude_attributes: this.excludeAttributes,
    };

    const result = await this.#fetch<SessionResponse>(
      "POST",
      "/sessions",
      body,
    );

    const sessionId = result.session_id;
    logger.debug(`Session initialized with ID: ${sessionId}`);
    return sessionId;
  }

  async #fetch<Result>(
    method: Http.Method,
    path: string,
    body?: unknown,
  ): Promise<Result> {
    const init: RequestInit = {
      method,
      signal: AbortSignal.timeout(HttpClient.TIMEOUT),
    };
    if (body != null) {
      init.headers = { "Content-Type": "application/json" };
      init.body = JSON.stringify(body);
    }

    const url = `${this.#baseUrl}/v1${path}`;

    const response = await fetch(url, init);

    if (!response.ok) {
      const errorText = await response.text();
      let detail = "";
      let stack = "";

      try {
        const errorData = ErrorResponse.parse(JSON.parse(errorText));
        detail = errorData.message;
        stack = `\n${errorData.stack}`;
      } catch (err) {
        logger.warn("Failed to parse error response as JSON: {err}{stack}", {
          err,
          stack,
        });
        detail = errorText;
      }
      throw new Error(
        `${init.method || "GET"} ${url} responded with ${response.status} ${response.statusText}: ${detail}`,
      );
    }

    return (await response.json()) as Result;
  }
}
